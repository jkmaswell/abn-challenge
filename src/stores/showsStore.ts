import { ShowModel } from '@/models/ShowModel'
import type { Show } from '@/types/show'
import { defineStore } from 'pinia'
import { computed, ref } from 'vue'

export const useTvShowsStore = defineStore('tvShows', () => {
  // State
  const shows = ref<Show[]>([])
  const searchQuery = ref('')

  // Actions
  const loadShows = async () => {
    const allShows = await ShowModel.fetchShows()

    shows.value = sortShowsByRating(allShows)
  }

  // Getters
  const genres = computed(() => {
    const allGenres = shows.value.flatMap(show => show.genres)
    
    return [...new Set(allGenres)]
  })

  const showsByGenre = computed(() => (genre: string) => sortShowsByRating(shows.value.filter(show => show.genres.includes(genre))))

  const filteredShows = computed(() => sortShowsByRating(
    shows.value.filter(show =>
      show.name.toLowerCase().includes(searchQuery.value.toLowerCase()),
    ),
  ))

  const sortShowsByRating = (shows: Show[]): Show[] => {
    return [...shows].sort((a, b) => (b.rating.average || 0) - (a.rating.average || 0))
  }

  return {
    shows,
    searchQuery,
    loadShows,
    genres,
    showsByGenre,
    filteredShows,
  }
})

export type TvShowsStore = ReturnType<typeof useTvShowsStore>;