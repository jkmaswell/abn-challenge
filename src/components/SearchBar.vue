<script setup lang="ts">
import { useTvShowsStore } from '@/stores/showsStore'
import { storeToRefs } from 'pinia'
import { ref, watch } from 'vue'

// Stores
const store = useTvShowsStore()
const { searchQuery } = storeToRefs(store)

// Data
const query = ref('')

// Watchers
watch(() => query.value, (nVal: string) => {
  searchQuery.value = nVal
})
</script>

<template>
  <div class="search-bar">
    <input
      v-model="query"
      type="text"
      placeholder="Search TV shows..."
      class="search-bar__input"
    >
  </div>
</template>

<style scoped lang="scss">
.search-bar {
  &__input {
    padding: 1rem;
    border: 0.1rem solid $border-color;
    border-radius: 0.4rem;
    font-size: 1.6rem;
    inline-size: 100%;
  }
}
</style>