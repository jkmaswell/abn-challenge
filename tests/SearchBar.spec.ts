import SearchBar from '@/components/SearchBar.vue'
import { useTvShowsStore } from '@/stores/showsStore'
import { mount } from '@vue/test-utils'
import { createPinia, setActivePinia } from 'pinia'
import { beforeEach, describe, expect, it, vi } from 'vitest'

vi.mock('@/store/tvShowsStore', () => ({
  useTvShowsStore: () => ({
    searchQuery: '',
  }),
}))

describe('SearchBar.vue', () => {
  beforeEach(() => {
    setActivePinia(createPinia())
  })

  it('renders input element correctly', () => {
    const wrapper = mount(SearchBar)
    const input = wrapper.find('input[type="text"]')

    expect(input.exists()).toBe(true)
  })

  it('binds input value to query', async () => {
    const store = useTvShowsStore()
    const wrapper = mount(SearchBar)
    const input = wrapper.find('input[type="text"]')
    
    await input.setValue('Breaking Bad')
    expect(wrapper.vm.query).toBe('Breaking Bad')
    expect(store.searchQuery).toBe('Breaking Bad')
  })

  it('updates store searchQuery when input value changes', async () => {
    const store = useTvShowsStore()
    const wrapper = mount(SearchBar)
    const input = wrapper.find('input[type="text"]')

    await input.setValue('Friends')
    expect(store.searchQuery).toBe('Friends')
  })

  it('matches snapshot', () => {
    const wrapper = mount(SearchBar)

    expect(wrapper.html()).toMatchSnapshot()
  })
})