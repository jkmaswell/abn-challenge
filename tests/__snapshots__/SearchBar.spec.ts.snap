// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar.vue > matches snapshot 1`] = `"<div data-v-6ab8b910="" class="search-bar"><input data-v-6ab8b910="" type="text" placeholder="Search TV shows..." class="search-bar__input"></div>"`;
